src/main/java/br/unb/cloudissues/http/JavaProjectsRetriever.java

```diff
@@ -1,218 +1,219 @@
+	private static final int DEFAULT_MAX_REQUESTS_TO_WAIT = 10;
+
+	private static final String PROJECTS_SEARCH_URL = "/components/search_projects";
+
+	private static int count = 1;
+
+	private final boolean isSonarCloud;
+
+	private final String baseUrl;
+
+	private final int timeToWaitInMiliSecondsBetweenRequests;
+	private final int maxRequestsToWait;
+
+	private final OkHttpClient httpClient = new OkHttpClient();
+
+	private JavaProjectsRetriever(Builder builder) {
+		baseUrl = builder.baseUrl;
+		isSonarCloud = Optional.ofNullable(builder.isSonarCloud).orElse(true);
+		timeToWaitInMiliSecondsBetweenRequests = Optional
+				.ofNullable(builder.timeToWaitInMiliSecondsBetweenRequests)
+				.orElse(DEFAULT_TIME_TO_WAIT_IN_MILISECONDS_BETWEEN_REQUESTS);
+		maxRequestsToWait = Optional.ofNullable(builder.maxRequestsToWait)
+				.orElse(DEFAULT_MAX_REQUESTS_TO_WAIT);
+	}
+
+	public static class Builder {
+
+		private final String baseUrl;
+		private Boolean isSonarCloud;
+		private Integer timeToWaitInMiliSecondsBetweenRequests;
+		private Integer maxRequestsToWait;
+
+		/**
+		 * @param baseUrl Sonar Web API Base URL until /api
+		 */
+		public Builder(String baseUrl) {
+			this.baseUrl = baseUrl;
+		}
+
+		public Builder isSonarCloud(boolean isSonarCloud) {
+			this.isSonarCloud = isSonarCloud;
+			return this;
+		}
+
+		public Builder withTimeToWaitInMiliSecondsBetweenRequests(int time) {
+			timeToWaitInMiliSecondsBetweenRequests = time;
+			return this;
+		}
+
+		public Builder withMaxRequestsToWait(int maxRequestsToWait) {
+			this.maxRequestsToWait = maxRequestsToWait;
+			return this;
+		}
+
+		public JavaProjectsRetriever build() {
+			return new JavaProjectsRetriever(this);
+		}
+	}
+
+	public List<Project> retrieve() throws IOException, InterruptedException {
+		Objects.requireNonNull(this.baseUrl);
+
+		count = 1;
+
+		System.out.println("\nretrieving Java projects in " + baseUrl + PROJECTS_SEARCH_URL);
+		return httpRequestJavaProjects();
+
+	}
+
+	private List<Project> httpRequestJavaProjects() throws IOException, InterruptedException {
+		String responseBody = retrieveResponseBodyForURL(buildURL());
+		Map<String, Object> responseMap = Utils.responseToMap(responseBody);
+		@SuppressWarnings("unchecked")
+		Long total = ((Map<String, Double>) responseMap.get("paging")).get("total").longValue();
+
+		List<Project> projects = new ArrayList<>(total.intValue());
+
+		List<SonarProjectComponent> projectsComponents = retrieveSonarProjectComponentsFromResponseMap(
+				responseMap);
+
+		projects.addAll(projectsComponents.stream().map(SonarProjectComponent::toProject)
+				.collect(toList()));
+
+		if (total > DEFAULT_PAGE_SIZE) {
+			projects.addAll(requestsProjectsForMoreThanOnePage(total));
+		}
+
+		return projects;
+	}
+
+	private String buildURL() {
+		HttpUrl.Builder urlBuilder = minimumUrlBuilder();
+		return urlBuilder.build().toString();
+	}
+
+	private HttpUrl.Builder minimumUrlBuilder() {
```

Comment:
#### Changes Here
    * Using the Object.requireNonNull because the operation Http.parse at any moment may returned null.
    * I think it's a great opportunity to show the user the fastet possible